torch/_inductor/codegen/cpp_flex_attention_template.py

```diff
@@ -0,0 +1,1093 @@
+
+        // Calculate scale * q @ k.T
+        auto i_kv = is_broadcast_bs_kv ? i/bs_shards : i;
+        auto j_kv = is_broadcast_head_kv ? j/gqa_shards : j;
+        auto kv_block_num = n / kvBlockSize;
+        auto kv_block_offset = n - kv_block_num * kvBlockSize;
+        // getting kv indices by [BS, Head, 1, kv_block_num]
+        auto i_kvi = is_broadcast_bs_kvi ? i/bs_shards_kvi : i;
+        auto j_kvi = is_broadcast_head_kvi ? j/gqa_shards_kvi : j;
+        auto kv_logical_data = kv_indices_data + i_kvi * kviStrideB +
+                                j_kvi * kviStrideH + kv_block_num;
+        if (!need_pack) {
+          auto k_addr_t = key_reorder_ptr + i * num_head * eheadSize * kvSize +
+                  j * eheadSize * kvSize + n * eheadSize;
+          at::native::cpublas::brgemm(
+              cur_qSplitSize,
+              cur_kvSplitSize,
+              eheadSize,
+              qStrideM,
+              cur_kvSplitSize,
+              cur_kvSplitSize,
+              false,
+              q_data + i * qStrideB + j * qStrideH +
+                  m * qStrideM,
+              k_addr_t,
+              qk_data,
+              need_pack);
+        } else {
+          at::native::cpublas::brgemm(
+              cur_qSplitSize,
+              cur_kvSplitSize,
+              eheadSize,
+              headSize_even ? qStrideM : eheadSize,
+              cur_kvSplitSize,
+              cur_kvSplitSize,
+              false,
+              !headSize_even
+                  ? query_t_padding_ptr
+                  : q_data + i * qStrideB + j * qStrideH + m * qStrideM,
+              key_reorder_ptr + i * num_head * eheadSize * kvSize +
+                  j * eheadSize * kvSize + n * eheadSize,
+              qk_data,
+              need_pack);
+        }
+
+        {{kernel.kernel_name}}_mul_scale_kernel<accum_t>(qk_data, scaling_factor, cur_qSplitSize*cur_kvSplitSize);
+
+{%- if score_mod and mask_mod %}
+        // apply score mod function
+        for (int64_t row = 0; row < cur_qSplitSize; ++row) {
+          for (int64_t col = 0; col < cur_kvSplitSize; col++) {
+            std::vector<int64_t> b_idx = {i};
+            std::vector<int64_t> h_idx = {j};
+            std::vector<int64_t> q_idx = {m+row};
+            int64_t phisical_kv_idx = n+col;
+            if (use_kv_indice) {
+                phisical_kv_idx= *kv_logical_data * kvBlockSize + col;
+            }
+            std::vector<int64_t> kv_idx = {phisical_kv_idx};
+            accum_t* in_ptr0 = qk_data + row * cur_kvSplitSize + col;
+            auto in_ptr1 = b_idx.data();
+            auto in_ptr2 = h_idx.data();
+            auto in_ptr3 = q_idx.data();
+            auto in_ptr4 = kv_idx.data();
+            {{ template.generate_other_buffer("score_others", 0, "len_score_other", kernel.args) }}
+            accum_t* out_ptr{{score_buf_idx}} = in_ptr0;
+            {{ template.modification(score_mod, score_buf_name, score_buf_idx) }}
```

Comment:
Is this vectorized? If not, add a TODO note here?

Reply:
not fully, and could be improved further, added TODO here, thanks.